[OSX] Replace FreeType by native glyph rasterization

michicc opened the pull request:
## Motivation / Problem

On OSX, text layout for TrueType fonts and more complex languages is done by CoreText. Font rendering on the other hand is done by the FreeType library, which assumes these two components never disagree.

Font selection in CoreText has lots of fallbacks built-in; it is extremly hard to _not_ get some font glyphs from CoreText. FreeType on the other hand needs a reference to an on-disk font file. We try hard to match the proper font, but might not always succeed if CoreText is applying some magic for exotic scripts. It is unlikely to ever be a problem for any kind of western script, though.

## Description

Letting CoreText render the font glyphs as well avoids any potential conflict or disagreement. It also removes an external dependency.

On average, CoreText appears to rasterize glyphs at least as fast as FreeType, often even faster. The implementation is similar to the Win32 implementation, as many concepts are analogue.

## Limitations

CoreText and FreeType make slightly different stylistic choices rasterizing glyphs. This is most apparent when anti-aliasing is activated, as CoreText has a tendency to render "thicker". CoreText appears to render best when using native (HiDPI) resolution (separate PR incoming).


## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* ~This PR affects the save game format? (label 'savegame upgrade')~
* ~This PR affects the GS/AI API? (label 'needs review: Script API')~
* ~This PR affects the NewGRF API? (label 'needs review: NewGRF')~
This is just a code move/rename, not a functionality change.

LordAro commented on src/fontcache.cpp:
```diff
@@ -1056,9 +1056,300 @@ static void LoadWin32Font(FontSize fs)
 	new Win32FontCache(fs, logfont, settings->size);
 }
 
+#elif defined(WITH_COCOA)
```
Think this file needs splitting into per-os variants. It's getting a bit silly

michicc replied:
Yes, except of course that e.g. FreeType is actually not a real OS-variant. Both OSX and Win will still use FreeType if built with it.

If you have a good idea how to organise the split (probably including fontdetection.cpp), I'm willing to do it.

orudge replied:
I suppose it would be per rendering engine, rather than per-OS, a bit like we have SDL, Windows, Mac video drivers.

I would be quite keen to get native font rendering in on Mac if we can, to avoid the freetype dependency. That said, I haven't yet tested this PR, but hope to do so when I get a chance.

TrueBrain commented on src/fontcache.cpp:
```diff
@@ -1056,9 +1056,301 @@ static void LoadWin32Font(FontSize fs)
+		pixels = scaled_height;
+
+		CFAutoRelease<CTFontRef> f(CTFontCreateWithFontDescriptor(this->font_desc.get(), 0.0f, nullptr));
+		if (f) {
```
`f`? What variable is that? :D Can I strongly suggest to use something not 1 letter? :D

TrueBrain commented on src/fontcache.cpp:
```diff
@@ -1056,9 +1056,301 @@ static void LoadWin32Font(FontSize fs)
+			float min_size = 0.0f;
+
+			/* The 'head' TrueType table contains information about the
+			 * 'smallest readable size in pixels'. Try to read it out. If
```
`read it out` <- I do not think this is correct English :D But that is a gut-feeling more than anything else :P

LordAro replied:
It's "bad" English, but I wouldn't say it's "incorrect" English :p
I would simply suggest "Try to read it, if ..."

TrueBrain commented on src/fontcache.cpp:
```diff
@@ -1056,9 +1056,301 @@ static void LoadWin32Font(FontSize fs)
+			if (data) {
+				uint16_t lowestRecPPEM; // At offset 46 of the 'head' TrueType table.
+				CFDataGetBytes(data.get(), CFRangeMake(46, sizeof(lowestRecPPEM)), (UInt8 *)&lowestRecPPEM);
+				min_size = CFSwapInt16BigToHost(lowestRecPPEM); // TrueType data is always big-endian.
```
I am sure you figured this one somehow .. maybe share the documentation link here? As `46` is a magic value .. maybe make a const out if it and point to the documentation there?

TrueBrain commented on src/fontcache.cpp:
```diff
@@ -1056,9 +1056,301 @@ static void LoadWin32Font(FontSize fs)
+			/* Font height is minimum height plus the difference between the default
+			 * height for this font size and the small size. */
+			int diff = scaled_height - ScaleFontTrad(_default_font_height[FS_SMALL]);
+			pixels = Clamp(std::min(min_size, 20.0f) + diff, scaled_height, MAX_FONT_SIZE);
```
`20.0f` <- magic value; please make a const out of it or comment where it comes from

TrueBrain commented on src/fontcache.cpp:
```diff
@@ -1056,9 +1056,301 @@ static void LoadWin32Font(FontSize fs)
+
+	this->font.reset(CTFontCreateWithFontDescriptor(this->font_desc.get(), pixels, nullptr));
+
+	/* Query the font metrics we needed. We generally round all values up to
```
`generally`, suggesting we don't always do it. Is this true?

TrueBrain commented on src/fontcache.cpp:
```diff
@@ -1056,9 +1056,301 @@ static void LoadWin32Font(FontSize fs)
+		return this->parent->MapCharToGlyph(key);
+	}
+
+	/* Convert characters outside of the BMP into surrogate pairs. */
```
`the BMP` .. which BMP?

TrueBrain commented on src/fontcache.cpp:
```diff
@@ -1056,9 +1056,301 @@ static void LoadWin32Font(FontSize fs)
+	CGGlyph glyph[2] = {0, 0};
+	if (CTFontGetGlyphsForCharacters(this->font.get(), chars, glyph, key >= 0x010000U ? 2 : 1)) {
+		return glyph[0];
+	} else {
```
You can drop the `else` here :)

TrueBrain commented on src/fontcache.cpp:
```diff
@@ -1056,9 +1056,301 @@ static void LoadWin32Font(FontSize fs)
+	/* Get glyph size. */
+	CGGlyph glyph = (CGGlyph)key;
+	CGRect bounds = CGRectNull;
+	if (MacOSVersionIsAtLeast(10, 8, 0)) {
```
Do we support 10.8? Weren't we 10.9+? (our description of the downloads suggests we are)

TrueBrain commented on src/fontcache.cpp:
```diff
@@ -1056,9 +1056,301 @@ static void LoadWin32Font(FontSize fs)
+	uint bb_height = (uint)std::ceil(bounds.size.height);
+
+	/* Add 1 pixel for the shadow on the medium font. Our sprite must be at least 1x1 pixel. */
+	uint width = std::max(1U, bb_width + (this->fs == FS_NORMAL));
```
`this->fs == FS_NORMAL ? 1 : 0` ? Implicit assuming true equals `1` always feels a bit iffy

michicc replied:
I am open to change it, but this is no assumption:

> "the type bool can be converted to int with the value false becoming ​0​ and true becoming 1."
> (https://en.cppreference.com/w/cpp/language/implicit_conversion#Numeric_promotions)

TrueBrain replied:
Sorry, I should have minded my words a bit better; I did not mean by specs or anything, as even for C we know that it will be fine. I meant from the reader perspective, it is an implicit statement, which means you have to notice it is a boolean, realise it is 0/1, etc.

TrueBrain commented on src/fontcache.cpp:
```diff
@@ -1056,9 +1056,301 @@ static void LoadWin32Font(FontSize fs)
+	uint height = std::max(1U, bb_height + (this->fs == FS_NORMAL));
+
+	/* Limit glyph size to prevent overflows later on. */
+	if (width > 256 || height > 256) usererror("Font glyph is too large");
```
Any chance we can constify these values? "No" is a valid answer

TrueBrain commented on src/fontcache.cpp:
```diff
@@ -1056,9 +1056,301 @@ static void LoadWin32Font(FontSize fs)
+		CGContextSetAllowsFontSubpixelQuantization(context.get(), !aa);
+		CGContextSetShouldSmoothFonts(context.get(), false);
+
+		float offset = aa ? 0.0f : 0.5f; // Align to pixel grid for non-AA.
```
aaaaaaaaaaaaaaa

Any chance on a better variable name? :D

TrueBrain commented on src/fontcache.cpp:
```diff
@@ -1056,9 +1056,301 @@ static void LoadWin32Font(FontSize fs)
+		CFRetain(font_ref.get()); // Increase ref count to match a later release.
+	}
+
+	if (!font_ref && MacOSVersionIsAtLeast(10, 6, 0)) {
```
Is this 10.6+ check needed, given we are 10.9+?

orudge replied:
People could still compile OpenTTD themselves on an older version, though 10.6 is rather old now admittedly. If it's new code, it's probably fine not to include such a check.

michicc replied:
My personal opinion here is that as long as backwards compatibility is simple (i.e. a single if like here), I try account for it.

TrueBrain commented on src/fontcache.cpp:
```diff
@@ -1056,9 +1056,301 @@ static void LoadWin32Font(FontSize fs)
+		new CoreTextFontCache(fs, std::move(font_ref), settings->size);
+	} else {
+		ShowInfoF("Unable to use '%s' for %s font, using sprite font instead", settings->font, SIZE_TO_NAME[fs]);
+	}
```
Can I suggest:

```
	if (!font_ref) {
		ShowInfoF("Unable to use '%s' for %s font, using sprite font instead", settings->font, SIZE_TO_NAME[fs]);
		return;
	}

	new CoreTextFontCache(fs, std::move(font_ref), settings->size);
```

?

TrueBrain reviewed (commented):
@orudge : did you have the chance to test this already? Would be good to have feedback from users on real mac hardware :)

orudge commented:
Finally got this tested again - the rendering looks higher quality than FreeType for sure. Particularly noticeable with bold fonts I've found (e.g., Tahoma Bold) - FreeType is very 'squashed together', with the native rendering looking considerably better. Certainly from a visual point of view (plus getting rid of a dependency in the Mac builds!) I'd be keen for this to go in.

michicc committed:
Codechange: Replace magic numbers by constants.

michicc committed:
Add: [OSX] Native font rendering without using FreeType.

michicc committed:
Change: [OSX] When auto-detecting a font, try for sans-serif first.

On a display, especially with small fonts or low pixel sizes, sans-serif
fonts are usually easier to read than serif fonts.

michicc commented:
@orudge Got a personal taste question for you:

If you enable font anti-aliasing (config option only I'm afraid), which rendering is better:

1. As in this PR.
2. If you change the 0.0f to a 0.5f as well at https://github.com/michicc/OpenTTD/blob/871bbb1fa7c1275157c56866cc629b2a0d12766e/src/os/macosx/font_osx.cpp#L306-L310

TrueBrain commented on src/os/macosx/font_osx.cpp:
```diff
@@ -136,3 +151,274 @@ bool SetFallbackFont(FreeTypeSettings *settings, const char *language_isocode, i
+	this->font.reset(CTFontCreateWithFontDescriptor(this->font_desc.get(), pixels, nullptr));
+
+	/* Query the font metrics we needed. We generally round all values up to
+	 * make sure we don't inadvertently cut off a row or column of pixels.,
```
nitpick: `.,` at end of sentence :D

TrueBrain commented on src/os/macosx/font_osx.cpp:
```diff
@@ -136,3 +151,274 @@ bool SetFallbackFont(FreeTypeSettings *settings, const char *language_isocode, i
+			 * 'smallest readable size in pixels'. Try to read it, if
+			 * that doesn't work, we use the default OS font size instead.
+			 *
+			 * Reference: https://developer.apple.com/fonts/TrueType-Reference-Manual/RM06/Chap6head.html */
```
really nice, tnx for this URL :D

TrueBrain reviewed (commented):
Looks really good; when you have an answer from @orudge, I see no blocking issues :)

orudge commented:
>     2. If you change the 0.0f to a 0.5f as well at https://github.com/michicc/OpenTTD/blob/871bbb1fa7c1275157c56866cc629b2a0d12766e/src/os/macosx/font_osx.cpp#L306-L310

This looks just a touch clearer to me I would say; as you say, it is very subtle.

michicc commented:
Artistic 2 choice applied.
